src/compiler/checker.ts

```diff
@@ -9318,6 +9777,86 @@ namespace ts {
                 pos < signature.parameters.length ? getTypeOfSymbol(signature.parameters[pos]) : anyType;
         }
 
+        // A function expression is a CommonJS Wrapper function if it has
+        // 1 parameter named 'require', two parameters named 'require' and 'exports',
+        // or 3 parameters named 'require', 'exports', and 'module' in that exact order
+        function isCommonJsWrapper(expression: FunctionExpression|ArrowFunction): boolean {
+           switch (expression.parameters.length) {
+                case 3:
+                    if ((<Identifier>expression.parameters[2].name).text !== "module") {
+                        return false;
+                    }
+                    // fall-through
+                case 2:
+                    if ((<Identifier>expression.parameters[1].name).text !== "exports") {
+                        return false;
+                    }
+                    // fall-through
+                case 1:
+                    if ((<Identifier>expression.parameters[0].name).text !== "require") {
+                        return false;
+                    }
+                    return true;
+                default:
+                    return false;
+            }
+        }
+
+        function assignDefineOrRequireCallParameterTypes(funcExpr: FunctionExpression|ArrowFunction) {

```

Comment:
As I mentioned earlier, this needs to be restructured so that the types are computed as contextual types.